scarb/src/core/manifest/target_defaults.rs

```diff
@@ -0,0 +1,103 @@
+use crate::core::TargetKind;
+use crate::core::{MaybeWorkspace, WorkspaceInherit};
+use anyhow::{Result, bail};
+use serde::{Deserialize, Serialize};
+use std::convert::TryFrom;
+use std::fmt;
+use typetag::serde;
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct TargetDefaults {
+    pub build_external_contracts: Vec<String>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct TomlTargetDefaults {
+    pub build_external_contracts: MaybeWorkspaceBuildExternalContracts,
+}
+
+impl From<TargetDefaults> for TomlTargetDefaults {
+    fn from(value: TargetDefaults) -> Self {
+        Self {
+            build_external_contracts: MaybeWorkspace::Defined(value.build_external_contracts),
+        }
+    }
+}
+
+pub type MaybeWorkspaceTargetDefaults =
+    MaybeWorkspace<TomlTargetDefaults, TomlWorkspaceTargetDefault>;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct TomlWorkspaceTargetDefault {
+    workspace: bool,
+}
+
+impl WorkspaceInherit for TomlWorkspaceTargetDefault {
+    fn inherit_toml_table(&self) -> &str {
+        "target-defaults"
+    }
+
+    fn workspace(&self) -> bool {
+        self.workspace
+    }
+}
+
+pub type MaybeWorkspaceBuildExternalContracts =
+    MaybeWorkspace<Vec<String>, TomlWorkspaceBuildExternalContracts>;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct TomlWorkspaceBuildExternalContracts {
+    workspace: bool,
+}
+
+impl WorkspaceInherit for TomlWorkspaceBuildExternalContracts {
+    fn inherit_toml_table(&self) -> &str {
+        "target-defaults.test"
+    }
+
+    fn workspace(&self) -> bool {
+        self.workspace
+    }
+}
+
+#[derive(Debug, Clone, Serialize, PartialEq, Eq, Ord, PartialOrd, Hash, Deserialize)]
+#[serde(try_from = "TargetKind")]
+pub struct TomlTargetKind(TargetKind);
```

Comment:
```suggestion
pub struct TomlTargetKindTestOnly(TargetKind);
```
I think this should be very explicit about the context in which it can be used.